mcs/class/Mono.Posix/Test/Mono.Unix.Native/SocketTest.cs

```diff
@@ -206,6 +206,651 @@ public void Shutdown ()
+				var actualAddress = new SockaddrIn ();
+				if (Syscall.getsockname (so1, actualAddress) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+				Assert.AreEqual (actualAddress.sa_family, UnixAddressFamily.AF_INET);
+				var port = Syscall.ntohs (actualAddress.sin_port);
+				Assert.IsTrue (port != 0);
+
+
+				var remoteAddress = new SockaddrIn {
+					sin_family = UnixAddressFamily.AF_INET,
+					sin_port = Syscall.htons (port),
+					sin_addr = new InAddr (127, 0, 0, 1),
+				};
+
+				// Send and receive a few bytes
+				long ret;
+				var buffer1 = new byte[] { 42, 43, 44 };
+				fixed (byte* ptr_buffer1 = buffer1) {
+					var iovecs1 = new Iovec[] {
+						new Iovec {
+							iov_base = (IntPtr) ptr_buffer1,
+							iov_len = (ulong) buffer1.Length,
+						},
+					};
+					var msghdr1 = new Msghdr {
+						msg_name = remoteAddress,
+						msg_iov = iovecs1,
+						msg_iovlen = 1,
+					};
+					ret = Syscall.sendmsg (so2, msghdr1, 0);
+					msghdr1.msg_name = remoteAddress.ToSockaddrStorage ();
+					if (ret >= 0)
+						ret = Syscall.sendmsg (so2, msghdr1, 0);
+				}
+				if (ret < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+
+				var senderAddress = new SockaddrIn ();
+				var senderAddressStorage = new SockaddrStorage ();
+				var buffer2 = new byte[1024];
+				var buffer3 = new byte[1024];
+				fixed (byte* ptr_buffer2 = buffer2, ptr_buffer3 = buffer3) {
+					var iovecs2 = new Iovec[] {
+						new Iovec {
+							iov_base = (IntPtr) ptr_buffer2,
+							iov_len = (ulong) buffer2.Length,
+						},
+					};
+					var msghdr2 = new Msghdr {
+						msg_name = senderAddress,
+						msg_iov = iovecs2,
+						msg_iovlen = 1,
+					};
+					ret = Syscall.recvmsg (so1, msghdr2, 0);
+					msghdr2.msg_name = senderAddressStorage;
+					iovecs2[0].iov_base = (IntPtr) ptr_buffer3;
+					if (ret >= 0)
+						ret = Syscall.recvmsg (so1, msghdr2, 0);
+				}
+				if (ret < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+				Assert.AreEqual (senderAddress.sa_family, UnixAddressFamily.AF_INET);
+				Assert.AreEqual (senderAddress.sin_addr, new InAddr (127, 0, 0, 1));
+				var senderAddress2 = SockaddrIn.FromSockaddrStorage (senderAddressStorage);
+				Assert.AreEqual (senderAddress2.sa_family, UnixAddressFamily.AF_INET);
+				Assert.AreEqual (senderAddress2.sin_addr, new InAddr (127, 0, 0, 1));
+
+				Assert.AreEqual (buffer1.Length, ret);
+				for (int i = 0; i < buffer1.Length; i++)
+					Assert.AreEqual (buffer1[i], buffer2[i]);
+				for (int i = 0; i < buffer1.Length; i++)
+					Assert.AreEqual (buffer1[i], buffer3[i]);
+			});
+		}
+
+		[Test]
+		public unsafe void ControlMsg ()
+		{
+			// Create two socket pairs and send inner_so1 and inner_so2 over the other socket pair using SCM_RIGHTS
+			WithSocketPair ((inner_so1, inner_so2) => {
+				WithSocketPair ((so1, so2) => {
+					// Create two SCM_RIGHTS control messages
+					var cmsg = new byte[2 * Syscall.CMSG_SPACE (sizeof (int))];
```

Comment:
> Not knowing what the value of Cmsghdr.Size, it is thus plausible (to idiots like me!) that Cmsghdr.Size is greater than what you actually allocated, new byte[2*Syscall.CMSG_SPACE(4)].

I don't think that's possible, as the space returned by CMSG_SPACE is supposed to contain a `struct cmsghdr`. Unfortunately, CMSG_SPACE does not seem to be part of POSIX (even though you need it to create a control message) :-$ However, it is described in [RFC 2292](https://tools.ietf.org/html/rfc2292#section-4.3.4) and it says "CMSG_SPACE() returns the space required by the object and its cmsghdr structure, including any padding needed to satisfy alignment requirements" which to me sounds like it should always return at least `sizeof (struct cmsghdr)`.

I think the reason for the union in the C code is simply make ensure the alignment is sufficient.

> I have to wonder if following the POSIX API too closely here is actually a good thing. This seems brittle as hell, easy to get wrong, not at all obvious, and I want to run screaming in terror. :-(

If you want I can try to add a high-level API for this. (Basically a `abstract class ControlMessage` + `class FDControlMessage : ControlMessage` and a way to convert between a `byte[]` and a `List<ControlMessage>`.)